File: LeetCode/Python/Hard/4037. Lexicographically Smallest Palindromic Permutation Greater Than Target/Solution.py
class Solution:
    def lexPalindromicPermutation(self, s: str, target: str) -> str:
        n = len(s)

        # Count characters
        cnt = [0] * 26
        for ch in s:
            cnt[ord(ch) - ord('a')] += 1

        # A palindrome can have at most one odd count
        odd = 0
        middle = ""

        for i in range(26):
            if cnt[i] % 2:
                odd += 1
                middle = chr(ord('a') + i)

        if odd > 1:
            return ""

        # For a palindrome, first half contains count // 2
        half_cnt = [x // 2 for x in cnt]
        half_len = n // 2

        def make_palindrome(half):
            return half + middle + half[::-1]

        # Build the smallest possible half
        half = []
        for i in range(26):
            half.extend([chr(ord('a') + i)] * half_cnt[i])

        half = ''.join(half)

        # If the smallest palindrome is already greater
        candidate = make_palindrome(half)
        if candidate > target:
            return candidate

        # We need the smallest half whose palindrome > target.
        # Since palindrome comparison is determined by the first half,
        # find the next permutation of half that is > target's first half.

        target_half = target[:half_len]

        # We need a half >= target_half.
        # Try to construct the smallest half greater than target_half.

        remaining = half_cnt[:]
        prefix = []

        # Match target_half as long as possible
        for ch in target_half:
            x = ord(ch) - ord('a')

            if remaining[x] > 0:
                prefix.append(ch)
                remaining[x] -= 1
            else:
                break

        # If we matched the whole target half,
        # the exact half may still produce a palindrome <= target.
        if len(prefix) == half_len:
            candidate = make_palindrome(''.join(prefix))

            if candidate > target:
                return candidate

        # Backtrack from the last matched position.
        for pos in range(len(prefix) - 1, -1, -1):

            # Return the character used at this position
            old = ord(prefix[pos]) - ord('a')
            remaining[old] += 1

            # Try the smallest character greater than it
            for c in range(old + 1, 26):

                if remaining[c] == 0:
                    continue

                new_half = prefix[:pos] + [chr(ord('a') + c)]

                remaining[c] -= 1

                # Append all remaining characters in sorted order
                for x in range(26):
                    new_half.extend(
                        [chr(ord('a') + x)] * remaining[x]
                    )

                half_string = ''.join(new_half)
                candidate = make_palindrome(half_string)

                if candidate > target:
                    return candidate

                remaining[c] += 1

        return ""

File: LeetCode/Java/Hard/4037. Lexicographically Smallest Palindromic Permutation Greater Than Target/Solution.java
class Solution {
    public String lexPalindromicPermutation(String s, String target) {
        int n = s.length();

        // Count characters
        int[] cnt = new int[26];

        for (char ch : s.toCharArray()) {
            cnt[ch - 'a']++;
        }

        // Check whether palindrome is possible
        int odd = 0;
        char middle = 0;

        for (int i = 0; i < 26; i++) {
            if (cnt[i] % 2 == 1) {
                odd++;
                middle = (char) ('a' + i);
            }
        }

        if (odd > 1) {
            return "";
        }

        int half = n / 2;

        // Count available characters for first half
        int[] halfCnt = new int[26];

        for (int i = 0; i < 26; i++) {
            halfCnt[i] = cnt[i] / 2;
        }

        // Try to make first half equal to target's first half
        int[] remaining = halfCnt.clone();
        boolean possible = true;

        for (int i = 0; i < half; i++) {
            int x = target.charAt(i) - 'a';

            if (remaining[x] == 0) {
                possible = false;
                break;
            }

            remaining[x]--;
        }

        if (possible) {
            String firstHalf = target.substring(0, half);

            // Even length
            if (n % 2 == 0) {
                String candidate = makePalindrome(firstHalf, "");
                if (candidate.compareTo(target) > 0) {
                    return candidate;
                }
            }

            // Odd length
            else {
                for (int c = 0; c < 26; c++) {
                    if (cnt[c] % 2 == 1) {
                        String mid = String.valueOf((char) ('a' + c));

                        String candidate =
                                makePalindrome(firstHalf, mid);

                        if (candidate.compareTo(target) > 0) {
                            return candidate;
                        }
                    }
                }
            }
        }

        // Need to make first half greater than target
        for (int i = half - 1; i >= 0; i--) {

            remaining = halfCnt.clone();

            boolean valid = true;

            // Match target[0 ... i-1]
            for (int j = 0; j < i; j++) {
                int x = target.charAt(j) - 'a';

                if (remaining[x] == 0) {
                    valid = false;
                    break;
                }

                remaining[x]--;
            }

            if (!valid) {
                continue;
            }

            // Choose smallest character greater than target[i]
            int start = target.charAt(i) - 'a' + 1;

            for (int c = start; c < 26; c++) {

                if (remaining[c] == 0) {
                    continue;
                }

                int[] rem = remaining.clone();
                rem[c]--;

                // Fill remaining positions with smallest characters
                StringBuilder firstHalf = new StringBuilder();

                firstHalf.append(target, 0, i);
                firstHalf.append((char) ('a' + c));

                for (int x = 0; x < 26; x++) {
                    while (rem[x] > 0) {
                        firstHalf.append((char) ('a' + x));
                        rem[x]--;
                    }
                }

                String mid = (n % 2 == 1)
                        ? String.valueOf(middle)
                        : "";

                String candidate =
                        makePalindrome(firstHalf.toString(), mid);

                if (candidate.compareTo(target) > 0) {
                    return candidate;
                }
            }
        }

        return "";
    }

    private String makePalindrome(String firstHalf, String middle) {
        StringBuilder result = new StringBuilder();

        result.append(firstHalf);
        result.append(middle);
        result.append(new StringBuilder(firstHalf).reverse());

        return result.toString();
    }
}